### PR TITLE
gs: Use Basic Station EUI as main identifier

### LINKED ISSUE #1076: Use Basic Station EUIs as main identifiers instead of IDs.
#### Summary

Use Basic Station EUIs as main identifiers instead of IDs.

#### Why do we need this?

Basic Station gateways are identified based on EUIs and not IDs and hence need to be handled similar to UDP.

Ref: [tti-issue](https://github.com/TheThingsIndustries/lorawan-stack/pull/1683)

#### What is already there? What do you see now?

The ID request URI is used as GatewayID for processing.

#### What is missing? What do you want to see?

The ID request URI should be converted to an EUI in before processing.

#### Environment

`ttn-lw-stack v3.1.0`

#### How do you propose to implement this?

Convert ID to EUI before processing.

#### Can you do this yourself and submit a Pull Request?

Yes

### PR DESCRIPTION
**Pull request by KrishnaIyer**
#### Summary

Closes #1076 

#### Changes

- Convert ID from request to EUI before usage.
- Adapt Auth check.

### CODE REVIEW

File: pkg/gatewayserver/io/basicstationlns/basicstationlns.go [hunk 1]
```diff
@@ -181,7 +189,19 @@ func (s *srv) handleTraffic(c echo.Context) error {
 		"remote_addr", c.Request().RemoteAddr,
 	))
 
-	ctx, ids, err := s.server.FillGatewayContext(ctx, ttnpb.GatewayIdentifiers{GatewayID: id})
+	// Convert the ID to EUI.
+	str := euiHexPattern.FindAllString(id, -1)
```
Comment (htdvisser) [inline pkg/gatewayserver/io/basicstationlns/basicstationlns.go]: Why not `FindString`?

File: pkg/gatewayserver/io/basicstationlns/basicstationlns.go [hunk 2]
```diff
@@ -181,7 +189,19 @@ func (s *srv) handleTraffic(c echo.Context) error {
...
+	if len(str) != 2 || len(str[1]) != 16 {
+		return errGatewayID.WithAttributes("id", id)
+	}
+	hexValue, err := hex.DecodeString(str[1])
+	if err != nil {
+		return errGatewayID.WithAttributes("id", id)
+	}
+	eui := types.EUI64{}
```
Comment (htdvisser) [inline pkg/gatewayserver/io/basicstationlns/basicstationlns.go]: I thought we preferred `var eui types.EUI64` in this situation? cc: @rvolosatovs?

File: pkg/gatewayserver/io/basicstationlns/basicstationlns.go [hunk 3]
```diff
@@ -152,25 +156,29 @@ func (s *srv) handleDiscover(c echo.Context) error {
 	return nil
 }
 
+var euiHexPattern = regexp.MustCompile("[a-f0-9A-F]+")
```
[the remark this answers is not in the document]
Comment (KrishnaIyer) [reply pkg/gatewayserver/io/basicstationlns/basicstationlns.go]: the request is like `eui-58a0cbfffe800cbf`. The first `e` in `eui` is a also a match so the length is checked once extracted.

File: pkg/gatewayserver/io/basicstationlns/basicstationlns.go (hunk 1 above)
Comment (KrishnaIyer) [reply pkg/gatewayserver/io/basicstationlns/basicstationlns.go]: Same as above. `FindString` exits on the first `e`.

File: pkg/gatewayserver/io/basicstationlns/basicstationlns.go (hunk 2 above)
Comment (KrishnaIyer) [reply pkg/gatewayserver/io/basicstationlns/basicstationlns.go]: Ok updated. I'm neutral on this.

File: pkg/gatewayserver/io/basicstationlns/basicstationlns.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (htdvisser) [reply pkg/gatewayserver/io/basicstationlns/basicstationlns.go]: Then shouldn't we change the regex to `^eui-([a-f0-9]{16})$`?

File: pkg/gatewayserver/io/basicstationlns/basicstationlns.go (hunk 1 above)
Comment (htdvisser) [reply pkg/gatewayserver/io/basicstationlns/basicstationlns.go]: If you change the regex as suggested above, this should become `FindStringSubmatch`

File: pkg/gatewayserver/io/basicstationlns/basicstationlns.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (KrishnaIyer) [reply pkg/gatewayserver/io/basicstationlns/basicstationlns.go]: That would also match `eui-58a0cbfffe800cbf`. The winner is  matching the substring with `[a-f0-9A-F]{16}`. 
(Note to self: Don't write regexes at the end of the day)

File: pkg/gatewayserver/io/basicstationlns/basicstationlns.go [hunk 4]
```diff
@@ -152,25 +156,29 @@ func (s *srv) handleDiscover(c echo.Context) error {
...
 }
 
+var euiHexPattern = regexp.MustCompile("[a-f0-9A-F]{16}")
+
 func (s *srv) handleTraffic(c echo.Context) error {
 	var sessionID int32
 	id := c.Param("id")
 	auth := c.Request().Header.Get(echo.HeaderAuthorization)
 	ctx := c.Request().Context()
+	md := metadata.MD{}
```
Comment (rvolosatovs) [inline pkg/gatewayserver/io/basicstationlns/basicstationlns.go]: If you apply https://github.com/TheThingsNetwork/lorawan-stack/pull/1077#discussion_r309272681 then it makes sense to do the same here.

**Comment by KrishnaIyer** [comment]
@htdvisser: Can you please merge this? I can't do it without @johanstokking's approval.

**Pull request merged**